### ISSUE TITLE
mwaa_export_data - Lambda: Invoke - Error - Notify_success

### ISSUE THREAD

**Issue by trsilva32** [aws-samples/amazon-mwaa-examples#61]
I'm getting the following error on the `notify_success` task:
"**botocore.errorfactory.TaskTimedOut: An error occurred (TaskTimedOut) when calling the SendTaskFailure operation: Task Timed Out: 'Provided task does not exist anymore'**"
Reading this [Stack Overflow](https://stackoverflow.com/questions/63850961/how-do-you-use-a-tasktoken-to-callback-to-the-same-step-function-twicel) solution it looks like the a taskToken can only be used once. 
Can anyone help me solve this problem, please?

Thanks

**Comment by skpune** [comment]
I am also facing the same error. Please let me know how to fix this issue.

**Comment by crupakheti** [comment]
Thank you, @trsilva32 and @skpune for reporting this issue. Could you please look at your DAG run metrics and report back the total time it took to reach to the notify_success task?

**Comment by pshen** [comment]
@trsilva32 I didn't face this problem. How did you pass the task token to Lambda function in the Step Functions' task?

**Comment by onisim-iacob** [comment]
Hi! @trsilva32 & @crupakheti,
 
I found a workaround, and changing this will stop working for 2.5.1 (and previous versions).

It happens if you use a MWAA >= 2.6.3. I managed to make it work with `export MWAA_ENV_VERSION=2.6.3` with dags from the [PR#59](https://github.com/aws-samples/amazon-mwaa-examples/pull/59#issue-1957863323).

The problem resides in the "Common" Lambda, which requires a tweak:

❗Broken block for 2.6.3 (around line 13.441 in index.js):
```
  async triggerDag(dagName, configuration) {
    const triggerResult = await this.execute(`dags trigger ${dagName}`, configuration);
    if (!triggerResult.stdOut.includes("triggered: True")) {
      throw new Error(`Dag [${dagName}] trigger failed with the following error: ${JSON.stringify(triggerResult)}`);
    }
    return triggerResult;
  }
```

✅ Fixed:

```
  async triggerDag(dagName, configuration) {
    const triggerResult = await this.execute(`dags trigger -o json ${dagName}`, configuration);
    if (!triggerResult.stdOut.includes('"external_trigger": "True"')) {
      throw new Error(`Dag [${dagName}] trigger failed with the following error: ${JSON.stringify(triggerResult)}`);
    }
    return triggerResult;
  }
```

The Lambda runs the `/aws_mwaa/cli`  call with `dags trigger` which requires an additional `-o json` and the condition for `"external_trigger": "True"`. 

Looking at this document between versions https://airflow.apache.org/docs/apache-airflow/2.6.3/cli-and-env-variables-ref.html#trigger shows that:

```
# 2.5.1
airflow dags trigger [-h] [-c CONF] [-e EXEC_DATE] [-r RUN_ID] [-S SUBDIR]
                     [-v]
                     dag_id
```

```
# 2.6.3
airflow dags trigger [-h] [-c CONF] [-e EXEC_DATE] [--no-replace-microseconds]
                     [-o table, json, yaml, plain] [-r RUN_ID] [-S SUBDIR]
                     [-v]
                     dag_id
```
From 2.6.3 adds extra `[-o table, json, yaml, plain]` parameter which is defaulted to `table`. This breaks the Lambda conditional and throws the _"botocore.errorfactory.TaskTimedOut: An error occurred (TaskTimedOut) when calling the SendTaskFailure operation: Task Timed Out: 'Provided task does not exist anymore'"_ error, which is misleading...

![image](https://github.com/aws-samples/amazon-mwaa-examples/assets/122359339/ba82accd-afcf-4939-9b0d-640ddfe91662)

By checking the Lambda log, I saw that the call indeed returns a table that breaks the "token" with this format.

![image](https://github.com/aws-samples/amazon-mwaa-examples/assets/122359339/08fd8619-c7cf-4d08-b478-6a810aadd420)

If you want to see what are all the possible outputs of this `-o` parameter looks like, take a look at this document to invoke the DAG locally: https://docs.aws.amazon.com/mwaa/latest/userguide/call-mwaa-apis-cli.html#create-cli-token-curl

```
CLI_JSON=$(aws mwaa --region YOUR_REGION create-cli-token --name YOUR_ENVIRONMENT_NAME) \
  && CLI_TOKEN=$(echo $CLI_JSON | jq -r '.CliToken') \
  && WEB_SERVER_HOSTNAME=$(echo $CLI_JSON | jq -r '.WebServerHostname') \
  && CLI_RESULTS=$(curl --request POST "https://$WEB_SERVER_HOSTNAME/aws_mwaa/cli" \
  --header "Authorization: Bearer $CLI_TOKEN" \
  --header "Content-Type: text/plain" \
  --data-raw "dags trigger YOUR_DAG_NAME") \
  && echo "Output:" \
  && echo $CLI_RESULTS | jq -r '.stdout' | base64 --decode \
  && echo "Errors:" \
  && echo $CLI_RESULTS | jq -r '.stderr' | base64 --decode
```

With this tests, I managed to debug the lambda call locally, and the problem resides on: `/aws_mwaa/cli` which behaves differently nowadays.



After adding the `-o json` parameter and the `"external_trigger": "True"` conditional, it works just like with the 2.5.1 😎👌🔥

![image](https://github.com/aws-samples/amazon-mwaa-examples/assets/122359339/4b333a4a-33b3-431d-a653-833a785e052c)

Take care!

**Comment by crupakheti** [comment]
Great find there, @onisim-iacob! Would you be able to raise a PR for us with your proposed changes?

**Comment by onisim-iacob** [comment]
@crupakheti here you have it [PR#67](https://github.com/aws-samples/amazon-mwaa-examples/pull/67)
Notice the warning in the PR: it will stop working for 2.5.1, but it is a good starting point to handle this problem :)

**Comment by crupakheti** [comment]
Fixed in PR #67.
